smarty/forcefield.py

```diff
@@ -1858,11 +1857,23 @@ def createForce(self, system, topology, nonbondedMethod=NoCutoff, nonbondedCutof
         # Set the partial charges based on reference molecules.
         for reference_molecule in topology._reference_molecules:
             atom_mappings = topology._reference_to_topology_atom_mappings[reference_molecule]
+            # Retrieve charges from reference molecule, stored by atom index
+            charge_by_atom = {}
+            for atom in reference_molecule.GetAtoms():
+                charge_by_atom[atom.GetIdx()] = atom.GetPartialCharge()
+
+            # Loop over mappings and copy NB parameters from reference molecule
+            # to other instances of the molecule
             for atom_mapping in atom_mappings:
-                for (atom, atom_index) in zip(reference_molecule.GetAtoms(), atom_mapping):
+                for (atom_index, map_atom_index) in atom_mapping.items():
+                    # Retrieve NB params for reference atom (charge not set yet)
                     [charge, sigma, epsilon] = force.getParticleParameters(atom_index)
-                    force.setParticleParameters(atom_index, atom.GetPartialCharge(), sigma, epsilon)
+                    # Look up the charge on the atom in the reference molecule
+                    charge = charge_by_atom[atom_index]*unit.elementary_charge
 
+                    # Set parameters for equivalent atom in other instance of
+                    # this molecule
+                    force.setParticleParameters(map_atom_index, charge, sigma, epsilon)
         # TODO: Should we check that there are no missing charges?
```

Comment:
Thanks. That's been there for a while, from John's initial draft of this code. At some point I may add a check that every molecule has charges.